### DISCUSSION TITLE
C++ integration: what is the most efficient way?

### DISCUSSION THREAD

**Discussion by panthavma** [bnpr/Malt#271] [Q&A]
Hello, I was wondering, what is the most efficient way to use C++ inside Malt pipelines?

I'm pretty sure we can just use some kind of python bindings, although that might not be the most efficient, considering we would have to pass the data twice (blender -> malt -> c++). Is there a better way to do so with Malt (or will there be one in the future) ?

**Comment by pragma37** [comment]
I guess it depends on what type of data you want to pass.
Most OpenGL objects are just handles, so you could pass them to a C++ DLL as-is using ctypes.
You can take a look at the internal CBlenderMalt and ipc libraries for reference.

If you mean higher level data like the scene, there's no built-in way. You could read the python objects from C++ using the cpython API or extract the data you need to a ctypes struct and pass that to your C++ functions.

For an eventual Malt 2.0, I would like to rewrite the core in C++ or Rust, and provide a C ABI compatible API, but at the moment almost everything is stored as Python objects.
